adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,253 @@
+            #
+            # Because of this, we prefer to use REQUEST_URI, if available, which is the
+            # relative path + query string, and doesn't mutate.
+            #
+            # REQUEST_URI is only available depending on what web server is running though.
+            # So when its not available, we want the original, unmutated PATH_INFO, which
+            # is just the relative path without query strings.
+            request_span_name = create_request_span_name(env['REQUEST_URI'] || original_env['PATH_INFO'])
+
+            # Sets as many attributes as are available before control
+            # is handed off to next middleware.
+            span = tracer.start_span(request_span_name,
+                                     with_parent_context: context,
+                                     # NOTE: try to set as many attributes via 'attributes' argument
+                                     #       instead of via span.set_attribute
+                                     attributes: request_span_attributes(env: env),
+                                     kind: :server)
+
+            context.set_value(current_span_key, span)
+          end
+
+          # return Context with the frontend span as the current span
+          def create_frontend_span(env, extracted_context)
+            # NOTE: get_request_start may return nil
+            request_start_time = OpenTelemetry::Adapters::Rack::Util::QueueTime.get_request_start(env)
+
+            return unless config[:record_frontend_span] && !request_start_time.nil?
+
+            # NOTE: start_span assumes context is managed explicitly,
+            #       while in_span and with_span activate span automatically
+            span = tracer.start_span('http_server.queue', # NOTE: span kind of 'proxy' is not defined
+                                     with_parent_context: extracted_context,
+                                     # NOTE: initialize with as many attributes as possible:
+                                     attributes: {
+                                       'component' => 'http',
+                                       'service' => config[:web_service_name],
+                                       'start_time' => request_start_time
+                                     },
+                                     kind: :server)
+
+            extracted_context.set_value(current_span_key, span)
+          end
+
+          def finish_span(context)
+            context[current_span_key]&.finish if context
+          end
+
+          def current_span_key
+            OpenTelemetry::Trace::Propagation::ContextKeys.current_span_key
+          end
+
+          def tracer
+            OpenTelemetry::Adapters::Rack::Adapter.instance.tracer
+          end
+
+          ### request_span
+
+          # Per https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#http-server-semantic-conventions
```

Comment:
This was super useful during development and review, although, I don't that this comment should hang around in the final version.